scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketSource.java

```diff
@@ -667,14 +671,26 @@ long getOrSampleByteSize() {
           .sum();
     }
 
-    KeyGroupIterator<byte[], V> createIterator(int bucketId, int targetParallelism) {
+    KeyGroupIterator<byte[], V> createIterator(
+        int bucketId, int targetParallelism, PipelineOptions options) {
+      final SortedBucketOptions opts = options.as(SortedBucketOptions.class);
+      final int bufferSize = opts.getReadBufferSize();
+      final boolean backoff = opts.getReadBackoff();
       final List<Iterator<V>> iterators =
           mapBucketFiles(
               bucketId,
               targetParallelism,
-              file -> {
+              (partition, file) -> {
                 try {
-                  return fileOperations.iterator(file);
+                  if (backoff) {
+                    // we hit 10,000ms limit when partition = 23, 32,442ms cumulatively
+                    // 26.71min at partition = 180
+                    long millis = (long) Math.min(10000, Math.pow(1.5, partition));
```

Comment:
info: so basically the delay would be proportional to the partition index which growth exponentially with the partition index. Did you check if a uniform/linear (i.e. proportional to the partition index) delay would result in a similar to exponential delay performance?

Reply:
Yeah they don't differ much since the 10000ms cap kicks in after 23 partitions. The only reason for an exponential increase is to not penalize jobs that reads reasonable number of partitions. Also according to the following doc, if we can spread such request spike over 20+ minutes that should give load balancer the time to ramp up.

> As a bucket approaches its IO capacity limit, Cloud Storage typically takes on the order of minutes to detect and accordingly redistribute the load across more servers.

> If your request rate is less than 1000 write requests per second or 5000 read requests per second, then no ramp-up is needed. If your request rate is expected to go over these thresholds, you should start with a request rate below or near the thresholds and then double the request rate no faster than every 20 minutes.

https://cloud.google.com/storage/docs/request-rate#redistribution